From c8cbf0e567790342d3e4f915a718966aa129a1be Mon Sep 17 00:00:00 2001
From: Srinivas Pandruvada <srinivas.pandruvada@linux.intel.com>
Date: Fri, 13 Jun 2014 13:09:47 -0700
Subject: [PATCH] iio: accel: kxcjk1013: standby mode after buffer activation

For several operations, the chip is expected to be in standby mode,
restore back the operation mode if the buffer is already enabled.
This is true for changing sampling rate and system resume operation.

Signed-off-by: Srinivas Pandruvada <srinivas.pandruvada@linux.intel.com>
---
 drivers/iio/accel/kxcjk-1013.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/iio/accel/kxcjk-1013.c b/drivers/iio/accel/kxcjk-1013.c
index a65e96a..ac85006 100644
--- a/drivers/iio/accel/kxcjk-1013.c
+++ b/drivers/iio/accel/kxcjk-1013.c
@@ -298,6 +298,9 @@ static int kxcjk1013_set_odr(struct kxcjk1013_data *data, int val, int val2)
 
 	data->odr_bits = odr_bits;
 
+	if (atomic_read(&data->power_state))
+		kxcjk1013_set_mode(data, OPERATION);
+
 	return 0;
 }
 
@@ -578,13 +581,14 @@ static int kxcjk1013_resume(struct device *dev)
 	struct kxcjk1013_data *data = iio_priv(indio_dev);
 
 	mutex_lock(&data->mutex);
-	kxcjk1013_set_mode(data, OPERATION);
+	if (atomic_read(&data->power_state))
+		kxcjk1013_set_mode(data, OPERATION);
 	mutex_unlock(&data->mutex);
 
 	return 0;
 }
 
-static SIMPLE_DEV_PM_OPS(kxcjk1013_pm_ops, kxcjk1013_suspend, NULL);
+static SIMPLE_DEV_PM_OPS(kxcjk1013_pm_ops, kxcjk1013_suspend, kxcjk1013_resume);
 #define KXCJK1013_PM_OPS (&kxcjk1013_pm_ops)
 #else
 #define KXCJK1013_PM_OPS NULL
-- 
2.0.0

